src/Extensions/TimedScopes/Internal/ActivityObserversIntializer.cs

```diff
@@ -14,6 +14,21 @@ namespace Microsoft.Omex.Extensions.TimedScopes
 {
 	internal sealed class ActivityObserversIntializer : IHostedService, IObserver<DiagnosticListener>, IObserver<KeyValuePair<string, object>>
 	{
+		private const string ActivityStartEnding = ".Start";
+
+		private const string ActivityStopEnding = ".Stop";
+
+		private static readonly string[] s_eventEndMarkersToListen = new[] {
```

Comment:
Is there any guaranteed format to these activities?

Reply:
Just thinking that, if it is always "Something.something.<EventEnd>", rather than enumerating the activity ends, and checking "EndsWith" for each relevant end; you could create a dictionary of <string, func> mapping relevant ends to operations.
You could then do a dictionary lookup on LastIndexOf(".") to end string length.

It's not the prettiest but it would save a lot of enumeration and since you've already called out sacrificing the use of Any for this reason, just said I'd suggest it.

Reply:
only source code as far as I know, but I'll try to find some documentation links

Reply:
Okay, I don't think this warrants blocking the review so I'm signing off.
Suggest making the improvement in another change, if it makes sense.

Reply:
https://github.com/aspnet/Hosting/blob/master/src/Microsoft.AspNetCore.Hosting/Internal/HostingApplicationDiagnostics.cs
Link to the code that produces it

Reply:
If you are really want to know what the fastest method is, you should check it with a timer like 
[Stopwatch](http://msdn.microsoft.com/en-us/library/system.diagnostics.stopwatch.aspx). I don't think there is a big difference in speed between Keiths and Andreys method. If the list of possible endings stays small.

Reply:
Sorry for some reason initially I've seen only your first message, and was answering on it

Reply:
I'll try to provide measurements

Reply:
I can bet that the code in the IL will be the same :). The compiler usually optimizes these details better than us.

Reply:
The most time is usually spent on reading the code, not executing it. Do whatever is easier to modify when needed.

Reply:
So it took longer then I've expected.
I've used the following benchmark:
```csharp
using BenchmarkDotNet.Attributes;
using BenchmarkDotNet.Jobs;
using BenchmarkDotNet.Running;
using System;
using System.Collections.Generic;
using System.Linq;

namespace PerfTest
{
  class Program
  {
    private static void Main(string[] args)
    {
      var summary = BenchmarkRunner.Run<IsEnableBenchmark>();
    }
  }

  [SimpleJob(RuntimeMoniker.Net472)]
  [SimpleJob(RuntimeMoniker.NetCoreApp31)]
  [MemoryDiagnoser]
  [MarkdownExporter]
  public class IsEnableBenchmark
  {
    private static readonly string[] array = new[] {
      ".Start",
      ".Stop",
      "RequestIn",
      "RequestOut",
    };

    private static readonly Dictionary<int,string[]> dictionary = new Dictionary<int, string[]>{
      { 6, new string[] { ".Start" } },
      { 5, new string[] { ".Stop" } },
      { 14, new string[] { ".HttpRequestIn" } },
      { 15, new string[] { ".HttpRequestOut" } },
    };

    private static bool EventEndsWith(string eventName, string ending) => eventName.EndsWith(ending, StringComparison.Ordinal);

    [Benchmark(Description = "Check using LINQ")]
    [ArgumentsSource(nameof(Data))]
    public bool IsEnabledLinq(string eventName) => 
      array.Any(ending => EventEndsWith(eventName, ending));

    [Benchmark(Baseline = true, Description = "Check using simple loop")]
    [ArgumentsSource(nameof(Data))]
    public bool IsEnabledList(string eventName)
    {
      foreach (string ending in array)
      {
        if (EventEndsWith(eventName, ending))
        {
          return true;
        }
      }

      return false;
    }

    [Benchmark(Description = "Check using dictionary for selecting list")]
    [ArgumentsSource(nameof(Data))]
    public bool IsEnabledDictionary(string eventName)
    {
      var index = eventName.Length - eventName.LastIndexOf('.');

      if (dictionary.TryGetValue(index, out var subList)) 
      {
        foreach (string ending in subList)
        {
          if (EventEndsWith(eventName, ending))
          {
            return true;
          }
        }
      }

      return false;
    }

    public IEnumerable<object[]> Data() => new[]
      {
        new object[] { "Microsoft.AspNetCore.Hosting.HttpRequestIn.Start" },
        new object[] { "Microsoft.AspNetCore.Hosting.HttpRequestIn.Stop" },
        new object[] { "System.Net.Http.HttpRequestOut.Start" },
        new object[] { "System.Net.Http.HttpRequestOut.Stop" },
        new object[] { "System.Net.Http.Exception" },
        new object[] { "System.Net.Http.HttpRequestOut" },
        new object[] { "Microsoft.AspNetCore.Hosting.HttpRequestIn" },
        new object[] { "Microsoft.ServiceFabric.Services.RemotingRequestOut" },
        new object[] { "Microsoft.ServiceFabric.Services.RemotingRequestIn" }
      };
  }
}

```
And I've got those results:
|                                      Method |       Runtime |            eventName |       Mean |     Error |     StdDev |     Median | Ratio | RatioSD |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|-------------------------------------------- |-------------- |--------------------- |-----------:|----------:|-----------:|-----------:|------:|--------:|-------:|------:|------:|----------:|
|                          'Check using LINQ' |    .NET 4.7.2 | Micro(...)estIn [42] | 105.176 ns | 2.1349 ns |  5.6985 ns | 104.626 ns |  3.13 |    0.23 | 0.0286 |     - |     - |     120 B |
|                   'Check using simple loop' |    .NET 4.7.2 | Micro(...)estIn [42] |  34.224 ns | 0.6409 ns |  0.5995 ns |  34.232 ns |  1.00 |    0.00 |      - |     - |     - |         - |
| 'Check using dictionary for selecting list' |    .NET 4.7.2 | Micro(...)estIn [42] |  39.690 ns | 0.3706 ns |  0.3094 ns |  39.583 ns |  1.16 |    0.02 |      - |     - |     - |         - |
|                                             |               |                      |            |           |            |            |       |         |        |       |       |           |
|                          'Check using LINQ' | .NET Core 3.1 | Micro(...)estIn [42] |  87.905 ns | 1.2554 ns |  1.0483 ns |  88.178 ns |  3.01 |    0.08 | 0.0286 |     - |     - |     120 B |
|                   'Check using simple loop' | .NET Core 3.1 | Micro(...)estIn [42] |  29.152 ns | 0.6085 ns |  0.7007 ns |  28.914 ns |  1.00 |    0.00 |      - |     - |     - |         - |
| 'Check using dictionary for selecting list' | .NET Core 3.1 | Micro(...)estIn [42] |  34.406 ns | 0.5760 ns |  0.5106 ns |  34.297 ns |  1.18 |    0.04 |      - |     - |     - |         - |
|                                             |               |                      |            |           |            |            |       |         |        |       |       |           |
|                          'Check using LINQ' |    .NET 4.7.2 | Micro(...)Start [48] |  56.236 ns | 0.9914 ns |  0.9273 ns |  56.443 ns |  3.75 |    0.12 | 0.0286 |     - |     - |     120 B |
|                   'Check using simple loop' |    .NET 4.7.2 | Micro(...)Start [48] |  14.991 ns | 0.3552 ns |  0.3323 ns |  14.870 ns |  1.00 |    0.00 |      - |     - |     - |         - |
| 'Check using dictionary for selecting list' |    .NET 4.7.2 | Micro(...)Start [48] |  34.649 ns | 0.7199 ns |  0.9611 ns |  34.631 ns |  2.34 |    0.07 |      - |     - |     - |         - |
|                                             |               |                      |            |           |            |            |       |         |        |       |       |           |
|                          'Check using LINQ' | .NET Core 3.1 | Micro(...)Start [48] |  51.778 ns | 1.7621 ns |  5.1401 ns |  50.680 ns |  5.07 |    0.55 | 0.0287 |     - |     - |     120 B |
|                   'Check using simple loop' | .NET Core 3.1 | Micro(...)Start [48] |   9.829 ns | 0.2758 ns |  0.3176 ns |   9.801 ns |  1.00 |    0.00 |      - |     - |     - |         - |
| 'Check using dictionary for selecting list' | .NET Core 3.1 | Micro(...)Start [48] |  27.573 ns | 0.5644 ns |  0.5543 ns |  27.661 ns |  2.78 |    0.08 |      - |     - |     - |         - |
|                                             |               |                      |            |           |            |            |       |         |        |       |       |           |
|                          'Check using LINQ' |    .NET 4.7.2 | Micro(...).Stop [47] |  81.432 ns | 1.6210 ns |  2.1078 ns |  81.426 ns |  3.39 |    0.12 | 0.0286 |     - |     - |     120 B |
|                   'Check using simple loop' |    .NET 4.7.2 | Micro(...).Stop [47] |  23.984 ns | 0.2766 ns |  0.2310 ns |  23.955 ns |  1.00 |    0.00 |      - |     - |     - |         - |
| 'Check using dictionary for selecting list' |    .NET 4.7.2 | Micro(...).Stop [47] |  34.202 ns | 0.7017 ns |  0.6891 ns |  33.896 ns |  1.43 |    0.04 |      - |     - |     - |         - |
|                                             |               |                      |            |           |            |            |       |         |        |       |       |           |
|                          'Check using LINQ' | .NET Core 3.1 | Micro(...).Stop [47] |  66.698 ns | 1.3685 ns |  3.6999 ns |  66.523 ns |  3.67 |    0.09 | 0.0286 |     - |     - |     120 B |
|                   'Check using simple loop' | .NET Core 3.1 | Micro(...).Stop [47] |  18.973 ns | 0.3295 ns |  0.3083 ns |  18.888 ns |  1.00 |    0.00 |      - |     - |     - |         - |
| 'Check using dictionary for selecting list' | .NET Core 3.1 | Micro(...).Stop [47] |  26.473 ns | 0.5587 ns |  0.7066 ns |  26.480 ns |  1.40 |    0.06 |      - |     - |     - |         - |
|                                             |               |                      |            |           |            |            |       |         |        |       |       |           |
|                          'Check using LINQ' |    .NET 4.7.2 | Micro(...)estIn [50] |  91.671 ns | 0.7718 ns |  0.6842 ns |  91.559 ns |  2.69 |    0.08 | 0.0286 |     - |     - |     120 B |
|                   'Check using simple loop' |    .NET 4.7.2 | Micro(...)estIn [50] |  34.281 ns | 0.7145 ns |  1.0472 ns |  33.837 ns |  1.00 |    0.00 |      - |     - |     - |         - |
| 'Check using dictionary for selecting list' |    .NET 4.7.2 | Micro(...)estIn [50] |  22.748 ns | 0.4810 ns |  0.7488 ns |  22.515 ns |  0.66 |    0.04 |      - |     - |     - |         - |
|                                             |               |                      |            |           |            |            |       |         |        |       |       |           |
|                          'Check using LINQ' | .NET Core 3.1 | Micro(...)estIn [50] |  83.766 ns | 1.7164 ns |  3.7675 ns |  82.949 ns |  2.91 |    0.11 | 0.0286 |     - |     - |     120 B |
|                   'Check using simple loop' | .NET Core 3.1 | Micro(...)estIn [50] |  29.085 ns | 0.6010 ns |  0.5902 ns |  29.090 ns |  1.00 |    0.00 |      - |     - |     - |         - |
| 'Check using dictionary for selecting list' | .NET Core 3.1 | Micro(...)estIn [50] |  18.350 ns | 0.4136 ns |  1.0299 ns |  18.061 ns |  0.66 |    0.03 |      - |     - |     - |         - |
|                                             |               |                      |            |           |            |            |       |         |        |       |       |           |
|                          'Check using LINQ' |    .NET 4.7.2 | Micro(...)stOut [51] | 116.585 ns | 2.3836 ns |  4.9226 ns | 116.851 ns |  2.58 |    0.12 | 0.0286 |     - |     - |     120 B |
|                   'Check using simple loop' |    .NET 4.7.2 | Micro(...)stOut [51] |  45.846 ns | 0.9014 ns |  0.9257 ns |  46.060 ns |  1.00 |    0.00 |      - |     - |     - |         - |
| 'Check using dictionary for selecting list' |    .NET 4.7.2 | Micro(...)stOut [51] |  24.728 ns | 0.5253 ns |  0.5621 ns |  24.553 ns |  0.54 |    0.02 |      - |     - |     - |         - |
|                                             |               |                      |            |           |            |            |       |         |        |       |       |           |
|                          'Check using LINQ' | .NET Core 3.1 | Micro(...)stOut [51] | 110.081 ns | 3.4112 ns | 10.0045 ns | 107.763 ns |  2.67 |    0.16 | 0.0286 |     - |     - |     120 B |
|                   'Check using simple loop' | .NET Core 3.1 | Micro(...)stOut [51] |  39.754 ns | 0.8096 ns |  0.8999 ns |  39.585 ns |  1.00 |    0.00 |      - |     - |     - |         - |
| 'Check using dictionary for selecting list' | .NET Core 3.1 | Micro(...)stOut [51] |  20.345 ns | 0.5325 ns |  1.5532 ns |  19.903 ns |  0.56 |    0.04 |      - |     - |     - |         - |
|                                             |               |                      |            |           |            |            |       |         |        |       |       |           |
|                          'Check using LINQ' |    .NET 4.7.2 | Syste(...)ption [25] | 114.792 ns | 2.4522 ns |  3.5169 ns | 114.598 ns |  2.79 |    0.20 | 0.0286 |     - |     - |     120 B |
|                   'Check using simple loop' |    .NET 4.7.2 | Syste(...)ption [25] |  40.978 ns | 0.8537 ns |  2.1883 ns |  40.628 ns |  1.00 |    0.00 |      - |     - |     - |         - |
| 'Check using dictionary for selecting list' |    .NET 4.7.2 | Syste(...)ption [25] |  17.338 ns | 0.1783 ns |  0.1489 ns |  17.336 ns |  0.43 |    0.02 |      - |     - |     - |         - |
|                                             |               |                      |            |           |            |            |       |         |        |       |       |           |
|                          'Check using LINQ' | .NET Core 3.1 | Syste(...)ption [25] | 100.903 ns | 2.0147 ns |  5.6494 ns | 101.416 ns |  2.78 |    0.16 | 0.0286 |     - |     - |     120 B |
|                   'Check using simple loop' | .NET Core 3.1 | Syste(...)ption [25] |  38.163 ns | 0.7892 ns |  1.2287 ns |  38.495 ns |  1.00 |    0.00 |      - |     - |     - |         - |
| 'Check using dictionary for selecting list' | .NET Core 3.1 | Syste(...)ption [25] |  14.165 ns | 0.3185 ns |  0.6787 ns |  14.089 ns |  0.37 |    0.02 |      - |     - |     - |         - |
|                                             |               |                      |            |           |            |            |       |         |        |       |       |           |
|                          'Check using LINQ' |    .NET 4.7.2 | Syste(...)stOut [30] | 117.770 ns | 3.3979 ns |  9.9118 ns | 114.863 ns |  2.94 |    0.28 | 0.0286 |     - |     - |     120 B |
|                   'Check using simple loop' |    .NET 4.7.2 | Syste(...)stOut [30] |  42.355 ns | 0.8790 ns |  1.3424 ns |  42.139 ns |  1.00 |    0.00 |      - |     - |     - |         - |
| 'Check using dictionary for selecting list' |    .NET 4.7.2 | Syste(...)stOut [30] |  43.173 ns | 0.8621 ns |  0.8064 ns |  42.860 ns |  1.01 |    0.05 |      - |     - |     - |         - |
|                                             |               |                      |            |           |            |            |       |         |        |       |       |           |
|                          'Check using LINQ' | .NET Core 3.1 | Syste(...)stOut [30] | 105.810 ns | 3.0751 ns |  9.0669 ns | 103.742 ns |  2.89 |    0.22 | 0.0286 |     - |     - |     120 B |
|                   'Check using simple loop' | .NET Core 3.1 | Syste(...)stOut [30] |  37.862 ns | 0.7590 ns |  0.9036 ns |  37.780 ns |  1.00 |    0.00 |      - |     - |     - |         - |
| 'Check using dictionary for selecting list' | .NET Core 3.1 | Syste(...)stOut [30] |  27.779 ns | 0.5137 ns |  0.4290 ns |  27.691 ns |  0.74 |    0.02 |      - |     - |     - |         - |
|                                             |               |                      |            |           |            |            |       |         |        |       |       |           |
|                          'Check using LINQ' |    .NET 4.7.2 | Syste(...)Start [36] |  66.191 ns | 2.3903 ns |  7.0104 ns |  65.424 ns |  4.35 |    0.48 | 0.0286 |     - |     - |     120 B |
|                   'Check using simple loop' |    .NET 4.7.2 | Syste(...)Start [36] |  15.863 ns | 0.4203 ns |  0.8490 ns |  15.598 ns |  1.00 |    0.00 |      - |     - |     - |         - |
| 'Check using dictionary for selecting list' |    .NET 4.7.2 | Syste(...)Start [36] |  36.450 ns | 0.7654 ns |  0.7517 ns |  36.207 ns |  2.29 |    0.13 |      - |     - |     - |         - |
|                                             |               |                      |            |           |            |            |       |         |        |       |       |           |
|                          'Check using LINQ' | .NET Core 3.1 | Syste(...)Start [36] |  52.165 ns | 1.6765 ns |  4.8905 ns |  51.605 ns |  5.54 |    0.46 | 0.0287 |     - |     - |     120 B |
|                   'Check using simple loop' | .NET Core 3.1 | Syste(...)Start [36] |   9.358 ns | 0.1808 ns |  0.1603 ns |   9.356 ns |  1.00 |    0.00 |      - |     - |     - |         - |
| 'Check using dictionary for selecting list' | .NET Core 3.1 | Syste(...)Start [36] |  24.171 ns | 0.6142 ns |  1.8014 ns |  23.483 ns |  2.89 |    0.13 |      - |     - |     - |         - |
|                                             |               |                      |            |           |            |            |       |         |        |       |       |           |
|                          'Check using LINQ' |    .NET 4.7.2 | Syste(...).Stop [35] |  68.683 ns | 1.3696 ns |  1.3451 ns |  68.530 ns |  3.09 |    0.07 | 0.0286 |     - |     - |     120 B |
|                   'Check using simple loop' |    .NET 4.7.2 | Syste(...).Stop [35] |  22.248 ns | 0.2147 ns |  0.1903 ns |  22.214 ns |  1.00 |    0.00 |      - |     - |     - |         - |
| 'Check using dictionary for selecting list' |    .NET 4.7.2 | Syste(...).Stop [35] |  31.060 ns | 0.4530 ns |  0.4016 ns |  31.018 ns |  1.40 |    0.03 |      - |     - |     - |         - |
|                                             |               |                      |            |           |            |            |       |         |        |       |       |           |
|                          'Check using LINQ' | .NET Core 3.1 | Syste(...).Stop [35] |  58.375 ns | 1.2031 ns |  2.9737 ns |  57.997 ns |  3.11 |    0.12 | 0.0286 |     - |     - |     120 B |
|                   'Check using simple loop' | .NET Core 3.1 | Syste(...).Stop [35] |  18.365 ns | 0.3790 ns |  0.3723 ns |  18.312 ns |  1.00 |    0.00 |      - |     - |     - |         - |
| 'Check using dictionary for selecting list' | .NET Core 3.1 | Syste(...).Stop [35] |  22.288 ns | 0.4767 ns |  0.5854 ns |  22.119 ns |  1.22 |    0.04 |      - |     - |     - |         - |

So it's really depends on use case, list solution better for events that's on top of the list and dictionary usage better on events that are not in the list.

I'm not sure what would be the distribution of events that we are getting, there would be definetly a lot of Start Stop events because we are producing them.

I'm inclined to leave it with the `foreach` solution because my main worry of moving for `Any` was memorry allocation during creation of closure that will contribute to memory trafic and GC cycles. And both `foreach` and `dictionary` not allocating anything. Beside that it's kindoff a compromise between complicated and performant solution.

One more side argument is that I'm intendend to add `Activity` creations during SF remoting that would be: `Microsoft.ServiceFabric.Services.RemotingRequestOut` and `Microsoft.ServiceFabric.Services.RemotingRequestIn`, if we will be using `foreach` approach it won't require any changes, same with any other `RequestOut` and `RequestIn`.

Reply:
Do you know if the benchmark class warms up the code before benchmarking? I would think it does.

Reply:
> 
> 
> The most time is usually spent on reading the code, not executing it. Do whatever is easier to modify when needed.

Except if it's inside 3 nested for loops in the critical path in the code :). By the way, "nested for loops go brrr": https://ericlippert.com/2020/03/27/new-grad-vs-senior-dev/

Reply:
Yes it does, you can run provided source code to get a full run although it takes a lot of time